# **Hacer solo un `runall` para evitar errores y revisar así!**

# 0. Utilidades

In [2]:
# Librerias
import pandas as pd
import re
from faker import Faker
from datetime import datetime

In [3]:
# Para inventar datos falsos:
fake = Faker()

In [4]:
# Carga de datos
books_df = pd.read_csv('GoodReads_Books.csv')
authors_df = pd.read_csv('GoodReads_Authors.csv')

In [5]:
books_df.head()

bookId  \
0                     2767052-the-hunger-games   
1  2.Harry_Potter_and_the_Order_of_the_Phoenix   
2                   2657.To_Kill_a_Mockingbird   
3                     1885.Pride_and_Prejudice   
4                               41865.Twilight   

                                       title                 series  \
0                           The Hunger Games    The Hunger Games #1   
1  Harry Potter and the Order of the Phoenix        Harry Potter #5   
2                      To Kill a Mockingbird  To Kill a Mockingbird   
3                        Pride and Prejudice                    NaN   
4                                   Twilight   The Twilight Saga #1   

                                      author  rating  \
0                            Suzanne Collins    4.33   
1  J.K. Rowling, Mary GrandPré (Illustrator)    4.50   
2                                 Harper Lee    4.28   
3  Jane Austen, Anna Quindlen (Introduction)    4.26   
4                            Stephenie Meyer    3.60   

                                         description language           isbn  \
0  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...  English  9780439023481   
1  There is a door at the end of a silent corrido...  English  9780439358071   
2  The unforgettable novel of a childhood in a sl...  English  9999999999999   
3  Alternate cover edition of ISBN 9780679783268S...  English  9999999999999   
4  About three things I was absolutely positive.\...  English  9780316015844   

                                              genres  \
0  ['Young Adult', 'Fiction', 'Dystopia', 'Fantas...   
1  ['Fantasy', 'Young Adult', 'Fiction', 'Magic',...   
2  ['Classics', 'Fiction', 'Historical Fiction', ...   
3  ['Classics', 'Fiction', 'Romance', 'Historical...   
4  ['Young Adult', 'Fantasy', 'Romance', 'Vampire...   

                                          characters  ... firstPublishDate  \
0  ['Katniss Everdeen', 'Peeta Mellark', 'Cato (H...  ...              NaN   
1  ['Sirius Black', 'Draco Malfoy', 'Ron Weasley'...  ...         06/21/03   
2  ['Scout Finch', 'Atticus Finch', 'Jem Finch', ...  ...         07/11/60   
3  ['Mr. Bennet', 'Mrs. Bennet', 'Jane Bennet', '...  ...         01/28/13   
4  ['Edward Cullen', 'Jacob Black', 'Laurent', 'R...  ...         10/05/05   

                                              awards numRatings  \
0  ['Locus Award Nominee for Best Young Adult Boo...    6376780   
1  ['Bram Stoker Award for Works for Young Reader...    2507623   
2  ['Pulitzer Prize for Fiction (1961)', 'Audie A...    4501075   
3                                                 []    2998241   
4  ['Georgia Peach Book Award (2007)', 'Buxtehude...    4964519   

                                      ratingsByStars likedPercent  \
0  ['3444695', '1921313', '745221', '171994', '93...         96.0   
1  ['1593642', '637516', '222366', '39573', '14526']         98.0   
2  ['2363896', '1333153', '573280', '149952', '80...         95.0   
3  ['1617567', '816659', '373311', '113934', '767...         94.0   
4  ['1751460', '1113682', '1008686', '542017', '5...         78.0   

                                             setting  \
0  ['District 12, Panem', 'Capitol, Panem', 'Pane...   
1  ['Hogwarts School of Witchcraft and Wizardry (...   
2               ['Maycomb, Alabama (United States)']   
3  ['United Kingdom', 'Derbyshire, England (Unite...   
4  ['Forks, Washington (United States)', 'Phoenix...   

                                            coverImg  bbeScore bbeVotes  price  
0  https://i.gr-assets.com/images/S/compressed.ph...   2993816    30516   5.09  
1  https://i.gr-assets.com/images/S/compressed.ph...   2632233    26923   7.38  
2  https://i.gr-assets.com/images/S/compressed.ph...   2269402    23328    NaN  
3  https://i.gr-assets.com/images/S/compressed.ph...   1983116    20452    NaN  
4  https://i.gr-assets.com/images/S/compressed.ph...   1459448    14874    2.1  

[5 rows x 25 columns]

# 1. Busqueda de autores comunes (creación de llave foránea)

In [6]:
def extract_primary_author(author_string):
    """
    Esta función extrae el primer nombre de autor principal del campo 'author'.
    Ignora coautores y otros adicionales como parentesis, mayusculas, etc.
    """
    # Eliminar todo lo que esté entre paréntesis, incluyendo los paréntesis
    author_string = re.sub(r'\(.*?\)', '', author_string)
    # Tomar solo el primer fragmento antes de una coma (si la hay)
    primary_author = author_string.split(",")[0]
    # Limpiar espacios extra y convertir todo a minúsculas
    primary_author = ' '.join(primary_author.strip().lower().split())
    return primary_author

In [7]:
# Aplica la función a la columna 'author' en el dataset de libros
books_df['author_clean'] = books_df['author'].apply(extract_primary_author)

# Aplica la función a la columna 'name' en el dataset de autores
authors_df['name_clean'] = authors_df['name'].str.lower()

# Reordenar las columnas en books_df
cols_books = books_df.columns.tolist()
cols_books.insert(cols_books.index('author') + 1, cols_books.pop(cols_books.index('author_clean')))
books_df = books_df[cols_books]

# Reordenar las columnas en authors_df
cols_authors = authors_df.columns.tolist()
cols_authors.insert(cols_authors.index('name') + 1, cols_authors.pop(cols_authors.index('name_clean')))
authors_df = authors_df[cols_authors]

In [8]:
# Autores Unicos:
unique_authors = authors_df[authors_df['name_clean'].isin(books_df['author_clean'].unique())]
len(unique_authors)

17918

# 2. Creación de las 3 Tablas

## Libro

| **Nombre Original (Hito 1)**            | **Nombre Simplificado (para implementación)** | **Tipo de Atributo**     |
|------------------------------------------|-----------------------------------------------|---------------------------|
| BookId                                   | `bookid`                                      | `String`                  |
| Título                                   | `titulo`                                      | `String`                  |
| Autor                                    | `autor`                                       | `String`                  |
| Descripción                              | `descripcion`                                 | `String`                  |
| Idioma                                   | `idioma`                                      | `String`                  |
| Géneros                                  | `generos`                                     | `Array de Strings`        |
| Personajes                               | `personajes`                                  | `Array de Strings`        |
| Formato del Libro                        | `formato`                                     | `String`                  |
| Páginas                                  | `paginas`                                     | `Int`                     |
| Editor                                   | `editor`                                      | `String`                  |
| Fecha de publicación                     | `fecha_pub`                                   | `String`                  |
| Primera fecha de Publicación             | `fecha_pub_inicial`                           | `String`                  |
| Premios                                  | `premios`                                     | `String`                  |
| Escenario                             | `ambientacion`                                | `String`                  |


In [9]:
# Creamos DataFrame Libro:
Libro0 = books_df.copy()

# Eliminamos atributos que no necesitamos:
columns_to_drop = ['series', 'author', 'rating', 'isbn', 'numRatings', 'ratingsByStars', 'likedPercent', 'coverImg', 'bbeScore', 'bbeVotes', 'price', 'edition']
Libro0 = Libro0.drop(columns=columns_to_drop)

# Renombramos los atributos a utilizar:
Libro0.rename(columns={
        'bookId': 'bookid',
        'title': 'titulo',
        'author_clean': 'autor',
        'description': 'descripcion',
        'language' : 'idioma',
        'genres': 'generos',
        'characters': 'personajes',
        'bookFormat': 'formato',
        'pages': 'paginas',
        'publisher': 'editor',
        'publishDate': 'fecha_pub',
        'firstPublishDate': 'fecha_pub_inicial',
        'awards': 'premios',
        'setting': 'ambientacion'
    }, inplace=True)

# Resultado final:
Libro0.head()

bookid  \
0                     2767052-the-hunger-games   
1  2.Harry_Potter_and_the_Order_of_the_Phoenix   
2                   2657.To_Kill_a_Mockingbird   
3                     1885.Pride_and_Prejudice   
4                               41865.Twilight   

                                      titulo            autor  \
0                           The Hunger Games  suzanne collins   
1  Harry Potter and the Order of the Phoenix     j.k. rowling   
2                      To Kill a Mockingbird       harper lee   
3                        Pride and Prejudice      jane austen   
4                                   Twilight  stephenie meyer   

                                         descripcion   idioma  \
0  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...  English   
1  There is a door at the end of a silent corrido...  English   
2  The unforgettable novel of a childhood in a sl...  English   
3  Alternate cover edition of ISBN 9780679783268S...  English   
4  About three things I was absolutely positive.\...  English   

                                             generos  \
0  ['Young Adult', 'Fiction', 'Dystopia', 'Fantas...   
1  ['Fantasy', 'Young Adult', 'Fiction', 'Magic',...   
2  ['Classics', 'Fiction', 'Historical Fiction', ...   
3  ['Classics', 'Fiction', 'Romance', 'Historical...   
4  ['Young Adult', 'Fantasy', 'Romance', 'Vampire...   

                                          personajes    formato paginas  \
0  ['Katniss Everdeen', 'Peeta Mellark', 'Cato (H...  Hardcover     374   
1  ['Sirius Black', 'Draco Malfoy', 'Ron Weasley'...  Paperback     870   
2  ['Scout Finch', 'Atticus Finch', 'Jem Finch', ...  Paperback     324   
3  ['Mr. Bennet', 'Mrs. Bennet', 'Jane Bennet', '...  Paperback     279   
4  ['Edward Cullen', 'Jacob Black', 'Laurent', 'R...  Paperback     501   

                             editor fecha_pub fecha_pub_inicial  \
0                  Scholastic Press  09/14/08               NaN   
1                   Scholastic Inc.  09/28/04          06/21/03   
2  Harper Perennial Modern Classics  05/23/06          07/11/60   
3                    Modern Library  10/10/00          01/28/13   
4         Little, Brown and Company  09/06/06          10/05/05   

                                             premios  \
0  ['Locus Award Nominee for Best Young Adult Boo...   
1  ['Bram Stoker Award for Works for Young Reader...   
2  ['Pulitzer Prize for Fiction (1961)', 'Audie A...   
3                                                 []   
4  ['Georgia Peach Book Award (2007)', 'Buxtehude...   

                                        ambientacion  
0  ['District 12, Panem', 'Capitol, Panem', 'Pane...  
1  ['Hogwarts School of Witchcraft and Wizardry (...  
2               ['Maycomb, Alabama (United States)']  
3  ['United Kingdom', 'Derbyshire, England (Unite...  
4  ['Forks, Washington (United States)', 'Phoenix...

## Valoracion

| **Nombre Original (Hito 1)**                     | **Nombre Simplificado (para la implementación)** | **Tipo de Atributo**     |
|--------------------------------------------------|--------------------------------------------------|---------------------------|
| BookId                                           | `bookid`                                         | `String`                  |
| Calificación                                     | `calificacion`                                   | `Float`                   |
| No de Calificaciones                             | `num_calificaciones`                              | `Int`                     |
| Calificaciones por Estrellas                     | `estrellas`                                      | `Array de Strings`        |
| Aprobación                                       | `porcentaje_likes`                                     | `Int`                   |
| Puntuación BBE                                  | `puntuacion_bbe`                                 | `Float`                   |
| BBE Votos                                        | `votos_bbe`                                      | `Int`                     |



In [10]:
# Creamos DataFrame Valoracion:
Valoracion0 = books_df.copy()

# Eliminamos atributos que no necesitamos:
columns_to_keep = ['bookId', 'rating', 'numRatings', 'ratingsByStars', 'likedPercent', 'bbeScore', 'bbeVotes']
Valoracion0 = Valoracion0[columns_to_keep]

# Renombramos los atributos a utilizar:
Valoracion0.rename(columns={
        'bookId': 'bookid',
        'rating': 'calificacion',
        'numRatings': 'num_calificaciones',
        'ratingsByStars': 'estrellas',
        'likedPercent': 'porcentaje_likes',
        'bbeScore': 'puntuacion_bbe',
        'bbeVotes': 'votos_bbe'
    }, inplace=True)

# Resultado final:
Valoracion0

bookid  calificacion  \
0                         2767052-the-hunger-games          4.33   
1      2.Harry_Potter_and_the_Order_of_the_Phoenix          4.50   
2                       2657.To_Kill_a_Mockingbird          4.28   
3                         1885.Pride_and_Prejudice          4.26   
4                                   41865.Twilight          3.60   
...                                            ...           ...   
52473                           11492014-fractured          4.00   
52474                             11836711-anasazi          4.19   
52475                              10815662-marked          3.70   
52476                         11330278-wayward-son          3.85   
52477                 10991547-daughter-of-helaman          4.02   

       num_calificaciones                                          estrellas  \
0                 6376780  ['3444695', '1921313', '745221', '171994', '93...   
1                 2507623  ['1593642', '637516', '222366', '39573', '14526']   
2                 4501075  ['2363896', '1333153', '573280', '149952', '80...   
3                 2998241  ['1617567', '816659', '373311', '113934', '767...   
4                 4964519  ['1751460', '1113682', '1008686', '542017', '5...   
...                   ...                                                ...   
52473                 871                  ['311', '310', '197', '42', '11']   
52474                  37                        ['16', '14', '5', '2', '0']   
52475                6674             ['2109', '1868', '1660', '647', '390']   
52476                 238                      ['77', '78', '59', '19', '5']   
52477                 246                     ['106', '73', '42', '17', '8']   

       porcentaje_likes  puntuacion_bbe  votos_bbe  
0                  96.0         2993816      30516  
1                  98.0         2632233      26923  
2                  95.0         2269402      23328  
3                  94.0         1983116      20452  
4                  78.0         1459448      14874  
...                 ...             ...        ...  
52473              94.0               0          1  
52474              95.0               0          1  
52475              84.0               0          1  
52476              90.0               0          1  
52477              90.0               0          1  

[52478 rows x 7 columns]

## Autor

| **Nombre Original (Hito 1)**                     | **Nombre Simplificado (para la implementación)** | **Tipo de Atributo**     |
|--------------------------------------------------|--------------------------------------------------|---------------------------|
| Nombre                                           | `nombre`                                         | `String`                  |
| Número de Trabajos                               | `num_trabajos`                                   | `Int`                     |
| Género                                           | `genero`                                         | `String`                  |
| Nacimiento                                       | `nacimiento`                                     | `String`                  |
| Muerte                                           | `muerte`                                         | `String`                  |
| Puntuación promedio                              | `puntuacion_prom`                            | `Float`                   |
| Número de votos                                   | `num_votos`                                      | `Int`                     |
| País                                             | `pais`                                           | `String`                  |


In [11]:
# Creamos DataFrame Autor:
Autor0 = unique_authors.copy()
Autor0.columns

# Mantener solo las columnas especificadas:
columns_to_keep = ['name_clean', 'workcount', 'gender', 'born', 'died', 'average_rate', 'rating_count', 'country']
Autor0 = Autor0[columns_to_keep]

# Renombramos los atributos a utilizar:
Autor0.rename(columns={
        'name_clean': 'nombre',
        'workcount': 'num_trabajos',
        'gender': 'genero',
        'born': 'nacimiento',
        'died': 'muerte',
        'average_rate': 'puntuacion_prom',
        'rating_count': 'num_votos',
        'country': 'pais'
    }, inplace=True)

# Resultado final
Autor0

nombre  num_trabajos   genero  nacimiento      muerte  \
24         lexy v. golden            30  unknown         NaN         NaN   
38        جهاد عبد الحميد             1  unknown         NaN         NaN   
51          keith roberts           118     male  1935-09-20  2000-10-05   
112        sean fay wolfe            25  unknown         NaN         NaN   
127          karren brady             6  unknown         NaN         NaN   
...                   ...           ...      ...         ...         ...   
209508  susan fraser king             4  unknown         NaN         NaN   
209509        manda scott            21   female         NaN         NaN   
209513    richard bandler            95  unknown         NaN         NaN   
209515       augusto cury            78  unknown  1958-10-02         NaN   
209516    paul pilkington            20     male         NaN         NaN   

        puntuacion_prom  num_votos            pais  
24                 4.05        514             NaN  
38                 3.52        591             NaN  
51                 3.69       5402  United Kingdom  
112                4.61       1737             NaN  
127                3.73        940             NaN  
...                 ...        ...             ...  
209508             3.78       6060             NaN  
209509             4.07      11936  United Kingdom  
209513             4.01       8053             NaN  
209515             3.82      10450           Macao  
209516             3.60      29803             NaN  

[17918 rows x 8 columns]

# 3. Limpieza, Creacion de Datos y de Tablas Finales

In [12]:
# Largos actuales de tablas:
print(len(Libro0), len(Valoracion0), len(Autor0))

52478 52478 17918


## 3.1 Libro

In [13]:
# Visualizacion de los datos
Libro0.head()

bookid  \
0                     2767052-the-hunger-games   
1  2.Harry_Potter_and_the_Order_of_the_Phoenix   
2                   2657.To_Kill_a_Mockingbird   
3                     1885.Pride_and_Prejudice   
4                               41865.Twilight   

                                      titulo            autor  \
0                           The Hunger Games  suzanne collins   
1  Harry Potter and the Order of the Phoenix     j.k. rowling   
2                      To Kill a Mockingbird       harper lee   
3                        Pride and Prejudice      jane austen   
4                                   Twilight  stephenie meyer   

                                         descripcion   idioma  \
0  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...  English   
1  There is a door at the end of a silent corrido...  English   
2  The unforgettable novel of a childhood in a sl...  English   
3  Alternate cover edition of ISBN 9780679783268S...  English   
4  About three things I was absolutely positive.\...  English   

                                             generos  \
0  ['Young Adult', 'Fiction', 'Dystopia', 'Fantas...   
1  ['Fantasy', 'Young Adult', 'Fiction', 'Magic',...   
2  ['Classics', 'Fiction', 'Historical Fiction', ...   
3  ['Classics', 'Fiction', 'Romance', 'Historical...   
4  ['Young Adult', 'Fantasy', 'Romance', 'Vampire...   

                                          personajes    formato paginas  \
0  ['Katniss Everdeen', 'Peeta Mellark', 'Cato (H...  Hardcover     374   
1  ['Sirius Black', 'Draco Malfoy', 'Ron Weasley'...  Paperback     870   
2  ['Scout Finch', 'Atticus Finch', 'Jem Finch', ...  Paperback     324   
3  ['Mr. Bennet', 'Mrs. Bennet', 'Jane Bennet', '...  Paperback     279   
4  ['Edward Cullen', 'Jacob Black', 'Laurent', 'R...  Paperback     501   

                             editor fecha_pub fecha_pub_inicial  \
0                  Scholastic Press  09/14/08               NaN   
1                   Scholastic Inc.  09/28/04          06/21/03   
2  Harper Perennial Modern Classics  05/23/06          07/11/60   
3                    Modern Library  10/10/00          01/28/13   
4         Little, Brown and Company  09/06/06          10/05/05   

                                             premios  \
0  ['Locus Award Nominee for Best Young Adult Boo...   
1  ['Bram Stoker Award for Works for Young Reader...   
2  ['Pulitzer Prize for Fiction (1961)', 'Audie A...   
3                                                 []   
4  ['Georgia Peach Book Award (2007)', 'Buxtehude...   

                                        ambientacion  
0  ['District 12, Panem', 'Capitol, Panem', 'Pane...  
1  ['Hogwarts School of Witchcraft and Wizardry (...  
2               ['Maycomb, Alabama (United States)']  
3  ['United Kingdom', 'Derbyshire, England (Unite...  
4  ['Forks, Washington (United States)', 'Phoenix...

In [14]:
# Contar valores nulos en cada columna del DataFrame Libro
null_counts = Libro0.isnull().sum()
print(null_counts)

bookid                   0
titulo                   0
autor                    0
descripcion           1338
idioma                3806
generos                  0
personajes               0
formato               1473
paginas               2347
editor                3696
fecha_pub              880
fecha_pub_inicial    21326
premios                  0
ambientacion             0
dtype: int64


In [15]:
# RELLENO CON DATOS FALSOS:

# Rellenar datos nulos en la columna 'descripcion' con descripciones falsas de 100 caracteres
Libro0['descripcion'] = Libro0['descripcion'].apply(lambda x: x if pd.notnull(x) else fake.text(max_nb_chars=100))

# Rellenar datos nulos en la columna 'idioma' con idiomas falsos
Libro0['idioma'] = Libro0['idioma'].apply(lambda x: x if pd.notnull(x) else fake.language_name())

# Rellenar datos nulos en la columna 'generos' con géneros falsos o cuando el dato es '[]':
Libro0['generos'] = Libro0['generos'].apply(lambda x: x if pd.notnull(x) and x != '[]' else fake.random_element(elements=('Fiction', 'Non-Fiction', 'Fantasy', 'Science Fiction', 'Mystery', 'Romance', 'Thriller', 'Horror', 'Biography', 'History', 'Self-Help', 'Cooking', 'Travel', 'Art', 'Science', 'Math', 'Poetry', 'Drama', 'Comics', 'Children', 'Young Adult', 'Adult')))

# Rellenar datos nulos en la columna 'personajes' con nombres de personajes falsos o cuando el dato es '[]':
Libro0['personajes'] = Libro0['personajes'].apply(lambda x: x if pd.notnull(x) and x != '[]' else fake.name())

# Rellenar datos nulos en la columna 'formato' con 'Hardcover' o 'Paperback'
Libro0['formato'] = Libro0['formato'].apply(lambda x: x if pd.notnull(x) else fake.random_element(elements=('Hardcover', 'Paperback')))

# Rellenar datos nulos en la columna 'paginas' con números aleatorios entre 50 y 1000
Libro0['paginas'] = Libro0['paginas'].apply(lambda x: x if pd.notnull(x) else fake.random_int(min=50, max=1000))

# Rellenar datos nulos en la columna 'editor' con nombres de editoriales falsas
Libro0['editor'] = Libro0['editor'].apply(lambda x: x if pd.notnull(x) else fake.company())

# Rellenar datos nulos en la columna 'fecha_pub' con fechas aleatorias entre 1900 y 2021 (en formato MM/DD/YY)
Libro0['fecha_pub'] = Libro0['fecha_pub'].apply(lambda x: x if pd.notnull(x) and not isinstance(x, str) else fake.date_between(start_date='-121y', end_date='today').strftime('%m/%d/%y'))

# Rellenar datos nulos en la columna 'fecha_pub_inicial' con fechas aleatorias entre 1900 y 2021 (en formato DD/MM/YY)
Libro0['fecha_pub_inicial'] = Libro0['fecha_pub_inicial'].apply(lambda x: x if pd.notnull(x) and not isinstance(x, str) else fake.date_between(start_date='-121y', end_date='today').strftime('%m/%d/%y'))

# Rellenar datos nulos en la columna 'premios' con nombres de premios falsos o cuando el dato es '[]':
Libro0['premios'] = Libro0['premios'].apply(lambda x: x if pd.notnull(x) and x != '[]' else fake.text(max_nb_chars=50))

# Rellenar datos nulos en la columna 'ambientacion' con nombres de lugares falsos o cuando el dato es '[]':
Libro0['ambientacion'] = Libro0['ambientacion'].apply(lambda x: x if pd.notnull(x) and x != '[]' else fake.city())


In [16]:
Libro0

bookid  \
0                         2767052-the-hunger-games   
1      2.Harry_Potter_and_the_Order_of_the_Phoenix   
2                       2657.To_Kill_a_Mockingbird   
3                         1885.Pride_and_Prejudice   
4                                   41865.Twilight   
...                                            ...   
52473                           11492014-fractured   
52474                             11836711-anasazi   
52475                              10815662-marked   
52476                         11330278-wayward-son   
52477                 10991547-daughter-of-helaman   

                                          titulo            autor  \
0                               The Hunger Games  suzanne collins   
1      Harry Potter and the Order of the Phoenix     j.k. rowling   
2                          To Kill a Mockingbird       harper lee   
3                            Pride and Prejudice      jane austen   
4                                       Twilight  stephenie meyer   
...                                          ...              ...   
52473                                  Fractured    cheri schmidt   
52474                                    Anasazi    emma michaels   
52475                                     Marked   kim richardson   
52476                                Wayward Son      tom pollack   
52477                        Daughter of Helaman     misty moncur   

                                             descripcion   idioma  \
0      WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...  English   
1      There is a door at the end of a silent corrido...  English   
2      The unforgettable novel of a childhood in a sl...  English   
3      Alternate cover edition of ISBN 9780679783268S...  English   
4      About three things I was absolutely positive.\...  English   
...                                                  ...      ...   
52473  The Fateful Trilogy continues with Fractured. ...  English   
52474  'Anasazi', sequel to 'The Thirteenth Chime' by...  English   
52475  --READERS FAVORITE AWARDS WINNER 2011--Sixteen...  English   
52476  A POWERFUL TREMOR UNEARTHS AN ANCIENT SECRETBu...  English   
52477  Fighting in Helaman's army is Keturah's deepes...  English   

                                                 generos  \
0      ['Young Adult', 'Fiction', 'Dystopia', 'Fantas...   
1      ['Fantasy', 'Young Adult', 'Fiction', 'Magic',...   
2      ['Classics', 'Fiction', 'Historical Fiction', ...   
3      ['Classics', 'Fiction', 'Romance', 'Historical...   
4      ['Young Adult', 'Fantasy', 'Romance', 'Vampire...   
...                                                  ...   
52473  ['Vampires', 'Paranormal', 'Young Adult', 'Rom...   
52474                         ['Mystery', 'Young Adult']   
52475  ['Fantasy', 'Young Adult', 'Paranormal', 'Ange...   
52476  ['Fiction', 'Mystery', 'Historical Fiction', '...   
52477  ['Lds Fiction', 'Historical Fiction', 'Young A...   

                                              personajes    formato paginas  \
0      ['Katniss Everdeen', 'Peeta Mellark', 'Cato (H...  Hardcover     374   
1      ['Sirius Black', 'Draco Malfoy', 'Ron Weasley'...  Paperback     870   
2      ['Scout Finch', 'Atticus Finch', 'Jem Finch', ...  Paperback     324   
3      ['Mr. Bennet', 'Mrs. Bennet', 'Jane Bennet', '...  Paperback     279   
4      ['Edward Cullen', 'Jacob Black', 'Laurent', 'R...  Paperback     501   
...                                                  ...        ...     ...   
52473                                    Anthony Charles       Nook       0   
52474                                      Brandon Combs  Paperback     190   
52475                                     Michael Taylor  Paperback     280   
52476                                        Cassie Lamb  Paperback     507   
52477                                      Jonathan Cook  Paperback     201   

                                 editor fecha_pub fecha_pub_inicial

In [17]:
# Contar valores nulos en cada columna del DataFrame Libro
null_counts = Libro0.isnull().sum()
print(null_counts)

bookid               0
titulo               0
autor                0
descripcion          0
idioma               0
generos              0
personajes           0
formato              0
paginas              0
editor               0
fecha_pub            0
fecha_pub_inicial    0
premios              0
ambientacion         0
dtype: int64


In [18]:
# Reemplazar valores no enteros en la columna 'paginas' con enteros aleatorios entre 50 y 1500
Libro0['paginas'] = Libro0['paginas'].apply(lambda x: x if isinstance(x, int) else fake.random_int(min=50, max=1500))

assert Libro0['paginas'].dtype == int

In [19]:
# Convertir la columna 'fecha_pub' a formato datetime donde los datos se encuentran en formato MM/DD/YY:
Libro0['fecha_pub'] = pd.to_datetime(Libro0['fecha_pub'], format='%m/%d/%y', errors='coerce')

# Extraer solo la parte de la fecha
Libro0['fecha_pub'] = Libro0['fecha_pub'].dt.date

assert Libro0['fecha_pub'].dtype == datetime

In [20]:
# Convertir la columna 'fecha_pub_inicial' a formato datetime donde los datos se encuentran en formato MM/DD/YY:
Libro0['fecha_pub_inicial'] = pd.to_datetime(Libro0['fecha_pub_inicial'], format='%m/%d/%y', errors='coerce')

# Extraer solo la parte de la fecha
Libro0['fecha_pub_inicial'] = Libro0['fecha_pub_inicial'].dt.date

assert Libro0['fecha_pub_inicial'].dtype == datetime

In [21]:
# UNICIDAD DE LA LLAVE BOOOKID EN LA TABLA LIBRO:

# Contar el número de duplicados que se eliminarán
num_duplicados = Libro0.duplicated(subset='bookid').sum()
print(f"Se eliminarán {num_duplicados} duplicados de la tabla Libro.")

# Eliminar filas duplicadas en la tabla Libro basadas en el bookid
Libro0 = Libro0.drop_duplicates(subset='bookid', keep='first')


# Verificar si hay duplicados
assert Libro0['bookid'].is_unique, 'Hay llaves bookid duplicadas en tabla Libro si falla el assert!'

Se eliminarán 54 duplicados de la tabla Libro.


In [22]:
# TABLA FINAL DE LIBRO:
Libro = Libro0.copy()
Libro

bookid  \
0                         2767052-the-hunger-games   
1      2.Harry_Potter_and_the_Order_of_the_Phoenix   
2                       2657.To_Kill_a_Mockingbird   
3                         1885.Pride_and_Prejudice   
4                                   41865.Twilight   
...                                            ...   
52473                           11492014-fractured   
52474                             11836711-anasazi   
52475                              10815662-marked   
52476                         11330278-wayward-son   
52477                 10991547-daughter-of-helaman   

                                          titulo            autor  \
0                               The Hunger Games  suzanne collins   
1      Harry Potter and the Order of the Phoenix     j.k. rowling   
2                          To Kill a Mockingbird       harper lee   
3                            Pride and Prejudice      jane austen   
4                                       Twilight  stephenie meyer   
...                                          ...              ...   
52473                                  Fractured    cheri schmidt   
52474                                    Anasazi    emma michaels   
52475                                     Marked   kim richardson   
52476                                Wayward Son      tom pollack   
52477                        Daughter of Helaman     misty moncur   

                                             descripcion   idioma  \
0      WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...  English   
1      There is a door at the end of a silent corrido...  English   
2      The unforgettable novel of a childhood in a sl...  English   
3      Alternate cover edition of ISBN 9780679783268S...  English   
4      About three things I was absolutely positive.\...  English   
...                                                  ...      ...   
52473  The Fateful Trilogy continues with Fractured. ...  English   
52474  'Anasazi', sequel to 'The Thirteenth Chime' by...  English   
52475  --READERS FAVORITE AWARDS WINNER 2011--Sixteen...  English   
52476  A POWERFUL TREMOR UNEARTHS AN ANCIENT SECRETBu...  English   
52477  Fighting in Helaman's army is Keturah's deepes...  English   

                                                 generos  \
0      ['Young Adult', 'Fiction', 'Dystopia', 'Fantas...   
1      ['Fantasy', 'Young Adult', 'Fiction', 'Magic',...   
2      ['Classics', 'Fiction', 'Historical Fiction', ...   
3      ['Classics', 'Fiction', 'Romance', 'Historical...   
4      ['Young Adult', 'Fantasy', 'Romance', 'Vampire...   
...                                                  ...   
52473  ['Vampires', 'Paranormal', 'Young Adult', 'Rom...   
52474                         ['Mystery', 'Young Adult']   
52475  ['Fantasy', 'Young Adult', 'Paranormal', 'Ange...   
52476  ['Fiction', 'Mystery', 'Historical Fiction', '...   
52477  ['Lds Fiction', 'Historical Fiction', 'Young A...   

                                              personajes    formato  paginas  \
0      ['Katniss Everdeen', 'Peeta Mellark', 'Cato (H...  Hardcover      291   
1      ['Sirius Black', 'Draco Malfoy', 'Ron Weasley'...  Paperback      577   
2      ['Scout Finch', 'Atticus Finch', 'Jem Finch', ...  Paperback      603   
3      ['Mr. Bennet', 'Mrs. Bennet', 'Jane Bennet', '...  Paperback      501   
4      ['Edward Cullen', 'Jacob Black', 'Laurent', 'R...  Paperback      169   
...                                                  ...        ...      ...   
52473                                    Anthony Charles       Nook     1097   
52474                                      Brandon Combs  Paperback      111   
52475                                     Michael Taylor  Paperback      963   
52476                                        Cassie Lamb  Paperback     1030   
52477                                      Jonathan Cook  Paperback      807   

                                 editor   fecha_pub fec

## 3.2 Valoracion

In [23]:
# Visualizacion de los datos
Valoracion0.head()

bookid  calificacion  \
0                     2767052-the-hunger-games          4.33   
1  2.Harry_Potter_and_the_Order_of_the_Phoenix          4.50   
2                   2657.To_Kill_a_Mockingbird          4.28   
3                     1885.Pride_and_Prejudice          4.26   
4                               41865.Twilight          3.60   

   num_calificaciones                                          estrellas  \
0             6376780  ['3444695', '1921313', '745221', '171994', '93...   
1             2507623  ['1593642', '637516', '222366', '39573', '14526']   
2             4501075  ['2363896', '1333153', '573280', '149952', '80...   
3             2998241  ['1617567', '816659', '373311', '113934', '767...   
4             4964519  ['1751460', '1113682', '1008686', '542017', '5...   

   porcentaje_likes  puntuacion_bbe  votos_bbe  
0              96.0         2993816      30516  
1              98.0         2632233      26923  
2              95.0         2269402      23328  
3              94.0         1983116      20452  
4              78.0         1459448      14874

In [24]:
# Contar valores nulos en cada columna del DataFrame Valoracion:
null_counts = Valoracion0.isnull().sum()
print(null_counts)

bookid                  0
calificacion            0
num_calificaciones      0
estrellas               0
porcentaje_likes      622
puntuacion_bbe          0
votos_bbe               0
dtype: int64


In [25]:
# RELLENO CON DATOS FALSOS:

# Rellenar datos nulos de porcentaje de likes con valores aleatorios entre 0 y 100:
Valoracion0['porcentaje_likes'] = Valoracion0['porcentaje_likes'].apply(lambda x: x if pd.notnull(x) else fake.random_int(min=0, max=100))

# Transformamos a Int los datos de la columna 'porcentaje_likes':
Valoracion0['porcentaje_likes'] = Valoracion0['porcentaje_likes'].astype(int)

In [26]:
# Contar valores nulos en cada columna del DataFrame Valoracion:
null_counts = Valoracion0.isnull().sum()
print(null_counts)

bookid                0
calificacion          0
num_calificaciones    0
estrellas             0
porcentaje_likes      0
puntuacion_bbe        0
votos_bbe             0
dtype: int64


In [27]:
# UNICIDAD DE LA LLAVE BOOKID EN VALORACION:

# Contar el número de duplicados que se eliminarán
num_duplicados = Valoracion0.duplicated(subset='bookid').sum()
print(f"Se eliminarán {num_duplicados} duplicados de la tabla Valoracion.")

# Eliminar filas duplicadas en la tabla Libro basadas en el bookid
Valoracion0 = Valoracion0.drop_duplicates(subset='bookid', keep='first')

# Verificar si hay duplicados
assert Valoracion0['bookid'].is_unique, 'Hay llaves bookid duplicadas en tabla Valoracion si falla el assert!'

Se eliminarán 54 duplicados de la tabla Valoracion.


In [28]:
# TABLA FINAL DE VALORACION:
Valoracion = Valoracion0.copy()
Valoracion.head()

bookid  calificacion  \
0                     2767052-the-hunger-games          4.33   
1  2.Harry_Potter_and_the_Order_of_the_Phoenix          4.50   
2                   2657.To_Kill_a_Mockingbird          4.28   
3                     1885.Pride_and_Prejudice          4.26   
4                               41865.Twilight          3.60   

   num_calificaciones                                          estrellas  \
0             6376780  ['3444695', '1921313', '745221', '171994', '93...   
1             2507623  ['1593642', '637516', '222366', '39573', '14526']   
2             4501075  ['2363896', '1333153', '573280', '149952', '80...   
3             2998241  ['1617567', '816659', '373311', '113934', '767...   
4             4964519  ['1751460', '1113682', '1008686', '542017', '5...   

   porcentaje_likes  puntuacion_bbe  votos_bbe  
0                96         2993816      30516  
1                98         2632233      26923  
2                95         2269402      23328  
3                94         1983116      20452  
4                78         1459448      14874

## 3.3 Autor

In [29]:
# Visualizacion de los datos
Autor0.head()

nombre  num_trabajos   genero  nacimiento      muerte  \
24    lexy v. golden            30  unknown         NaN         NaN   
38   جهاد عبد الحميد             1  unknown         NaN         NaN   
51     keith roberts           118     male  1935-09-20  2000-10-05   
112   sean fay wolfe            25  unknown         NaN         NaN   
127     karren brady             6  unknown         NaN         NaN   

     puntuacion_prom  num_votos            pais  
24              4.05        514             NaN  
38              3.52        591             NaN  
51              3.69       5402  United Kingdom  
112             4.61       1737             NaN  
127             3.73        940             NaN

In [30]:
# Contar valores nulos en cada columna del DataFrame Autor:
null_counts = Autor0.isnull().sum()
print(null_counts)

nombre                 0
num_trabajos           0
genero                 0
nacimiento         10671
muerte             14589
puntuacion_prom        0
num_votos              0
pais                8899
dtype: int64


In [31]:
# RELLENO CON DATOS FALSOS:

# Rellenar datos nulos en la columna 'genero' con 'male' o 'female' si el genero es 'unknown'
Autor0['genero'] = Autor0['genero'].apply(lambda x: x if x != 'unknown' else fake.random_element(elements=('male', 'female')))

# Rellenar datos nulos en la columna 'nacimiento' con fechas aleatorias entre 1800 y 2002 (en formato YYYY-MM-DD)
Autor0['nacimiento'] = Autor0['nacimiento'].apply(lambda x: x if pd.notnull(x) else fake.date_of_birth(minimum_age=18, maximum_age=100).strftime('%Y-%m-%d'))

# Convertir la columna 'nacimiento' a tipo datetime:
Autor0['nacimiento'] = pd.to_datetime(Autor0['nacimiento'])

# Rellenar datos nulos en la columna 'muerte' con fechas aleatorias entre 1900 y 2021 (en formato YYYY-MM-DD) tal que la persona tenga almenos 18 años antes de morir:
Autor0['muerte'] = Autor0.apply(lambda row: row['muerte'] if pd.notnull(row['muerte']) else fake.date_between(start_date=row['nacimiento'] + pd.DateOffset(years=18), end_date='today').strftime('%Y-%m-%d'), axis=1)

# Convertir la columna 'muerte' a tipo datetime:
Autor0['muerte'] = pd.to_datetime(Autor0['muerte'])

# Rellenar la columna 'pais' con nombres de paises falsos:
Autor0['pais'] = Autor0['pais'].apply(lambda x: x if pd.notnull(x) else fake.country())

In [32]:
# Contar valores nulos en cada columna del DataFrame Autor:
null_counts = Autor0.isnull().sum()
print(null_counts)

nombre             0
num_trabajos       0
genero             0
nacimiento         0
muerte             0
puntuacion_prom    0
num_votos          0
pais               0
dtype: int64


In [33]:
# Ajusta el indice de Autor0 para que parta en 0:
Autor0.reset_index(drop=True, inplace=True)

In [34]:
# TABLA FINAL DE AUTOR:
Autor = Autor0.copy()
Autor.head()

nombre  num_trabajos  genero nacimiento     muerte  \
0   lexy v. golden            30  female 1938-10-11 2013-02-14   
1  جهاد عبد الحميد             1    male 1934-05-06 1982-04-01   
2    keith roberts           118    male 1935-09-20 2000-10-05   
3   sean fay wolfe            25    male 1958-08-23 2017-10-26   
4     karren brady             6  female 1930-08-05 1970-05-16   

   puntuacion_prom  num_votos            pais  
0             4.05        514      Montenegro  
1             3.52        591         Eritrea  
2             3.69       5402  United Kingdom  
3             4.61       1737         Hungary  
4             3.73        940       Hong Kong

# 4. Recorte Final de Filas (eliminación de casos problemáticos)

In [35]:
# Largos actuales de tablas:
print(len(Libro), len(Valoracion), len(Autor))

52424 52424 17918


In [36]:
# Verifica que todos los autores en la tabla Libro estén en la tabla Autor (para efectividad de la llave foránea)
autores_faltantes = Libro[~Libro['autor'].isin(Autor['nombre'])]['autor']

# Imprimir los autores que faltan
if not autores_faltantes.empty:
    print(f"{len(autores_faltantes)} no se encuentran en la tabla Autor:")
    print(autores_faltantes.tolist())
else:
    print("Todos los autores en la tabla Libro están presentes en la tabla Autor.")

6982 no se encuentran en la tabla Autor:
['richard adams', 'arthur miller', 'thomas harris', 'سید مرتضی مصطفوی', 'arthur miller', 'kyle keyes', 'diana forbes', 'abraham verghese', 'diamante lavendar', 'kyle keyes', 'kyle keyes', 'suzanne boisvert', 'سید مرتضی مصطفوی', 'hubert seljr.', 'aimee cabo nikolov', 'c. toni graham', 'traci medford-rosow', 'kenneth thomas', 'sergio cobo', 'peter b. forster', 'kathleen zamboni mccormick', 'james mcbride', 'vera jane cook', 'thomas harris', 'kenneth jarrett singleton', 'ammar habib', 'karen glista', 'andrea luhman', 'trevor alan foris', 'm. robinson', 'maria nhambu', 'fred misurella', 'anne willow', 'kenneth thomas', 'kyle keyes', 'vincent panettiere', 'mary stewart', 'd.s. smith', 'm. robinson', 'omar farhad', 'afroz alam', 'michael g. kramer', 'michel guyon', 'yuehai xiao', 'ammar habib', 'rohith s. katbamna', 'simon okill', 'karen prince', 'c. toni graham', 'gillian dance', 'rebecca donovan', 'gabriel f.w. koch', 'سید مرتضی مصطفوی', 'maria nham

In [37]:
# Identificar los IDs de los libros asociados a los autores faltantes para eliminarlos de la tabla Libro

libros_faltantes_ids = Libro[Libro['autor'].isin(autores_faltantes)]['bookid']

assert len(libros_faltantes_ids) == len(autores_faltantes)

In [38]:
# Eliminar las filas de la tabla Valoracion cuyos bookid están en libros_faltantes_ids
Valoracion = Valoracion[~Valoracion['bookid'].isin(libros_faltantes_ids)]

# Eliminar los libros de la tabla Libro cuyos autores están en autores_faltantes
Libro = Libro[~Libro['autor'].isin(autores_faltantes)]

assert len(Valoracion) == len(Libro)

In [39]:
duplicados = Autor[Autor.duplicated(subset=['nombre'], keep=False)] #hay 4 nombres repetidos. Se eliminaran las 8 filas duplicadas. al ejecutarlo denuevo tras un runall deberia dar vacio.
duplicados

nombre  num_trabajos  genero nacimiento     muerte  \
4979            hesiod           102  female 1958-11-14 1983-02-22   
5826    miroslav antić             2    male 2003-10-29 2024-02-17   
5827    miroslav antić            25    male 1932-03-14 1986-06-24   
6923            hesiod             1    male 1931-08-28 1955-06-26   
7131      mike mignola             1  female 1931-03-27 1968-05-08   
11451  joshua e. hanft             9  female 1979-11-23 2011-06-01   
13271     mike mignola          1011    male 1960-09-16 2004-06-08   
17287  joshua e. hanft             1    male 2006-01-07 2024-06-28   

       puntuacion_prom  num_votos                  pais  
4979              3.78      19039                Greece  
5826              4.66        104  Syrian Arab Republic  
5827              4.59       2085                Serbia  
6923              3.47        180              Zimbabwe  
7131              3.78        469            San Marino  
11451             3.88        390               Vietnam  
13271             4.11     321613         United States  
17287             4.14        950    Dominican Republic

In [40]:
# ELIMINAR LLAVES DUPLICADAS EN Autor:

# Eliminar duplicados de la tabla Autor, eliminando las dos filas duplicadas:
Autor = Autor.drop(duplicados.index)

# Eliminar las filas en la tabla Libro cuyos autores están en la lista de duplicados
Libro = Libro[~Libro['autor'].isin(duplicados['nombre'])]

# Eliminar las filas en la tabla Valoracion cuyos bookid están en la tabla Libro
Valoracion = Valoracion[Valoracion['bookid'].isin(Libro['bookid'])]

# Verificar los resultados
assert len(Libro) == len(Valoracion)
assert len(Autor) == 17910 # (17918 - 8)

print(len(Libro), len(Valoracion), len(Autor))

45429 45429 17910


In [41]:
# VERIFICACION FINAL DE VALIDEZ DE LA LLAVE:

# Verifica que todos los autores en la tabla Libro estén en la tabla Autor
autores_faltantes = Libro[~Libro['autor'].isin(Autor['nombre'])]['autor']

# Imprimir los autores que faltan
if not autores_faltantes.empty:
    print(f"{len(autores_faltantes)} no se encuentran en la tabla Autor:")
    print(autores_faltantes.tolist())
else:
    print("Todos los autores en la tabla Libro están presentes en la tabla Autor. GG!")

Todos los autores en la tabla Libro están presentes en la tabla Autor. GG!


# 5. Visualizacion de Tablas Finales (Libro, Valoracion y Autor)

In [42]:
Libro.head()

bookid  \
0                     2767052-the-hunger-games   
1  2.Harry_Potter_and_the_Order_of_the_Phoenix   
2                   2657.To_Kill_a_Mockingbird   
3                     1885.Pride_and_Prejudice   
4                               41865.Twilight   

                                      titulo            autor  \
0                           The Hunger Games  suzanne collins   
1  Harry Potter and the Order of the Phoenix     j.k. rowling   
2                      To Kill a Mockingbird       harper lee   
3                        Pride and Prejudice      jane austen   
4                                   Twilight  stephenie meyer   

                                         descripcion   idioma  \
0  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...  English   
1  There is a door at the end of a silent corrido...  English   
2  The unforgettable novel of a childhood in a sl...  English   
3  Alternate cover edition of ISBN 9780679783268S...  English   
4  About three things I was absolutely positive.\...  English   

                                             generos  \
0  ['Young Adult', 'Fiction', 'Dystopia', 'Fantas...   
1  ['Fantasy', 'Young Adult', 'Fiction', 'Magic',...   
2  ['Classics', 'Fiction', 'Historical Fiction', ...   
3  ['Classics', 'Fiction', 'Romance', 'Historical...   
4  ['Young Adult', 'Fantasy', 'Romance', 'Vampire...   

                                          personajes    formato  paginas  \
0  ['Katniss Everdeen', 'Peeta Mellark', 'Cato (H...  Hardcover      291   
1  ['Sirius Black', 'Draco Malfoy', 'Ron Weasley'...  Paperback      577   
2  ['Scout Finch', 'Atticus Finch', 'Jem Finch', ...  Paperback      603   
3  ['Mr. Bennet', 'Mrs. Bennet', 'Jane Bennet', '...  Paperback      501   
4  ['Edward Cullen', 'Jacob Black', 'Laurent', 'R...  Paperback      169   

                             editor   fecha_pub fecha_pub_inicial  \
0                  Scholastic Press  2012-10-08        1990-10-17   
1                   Scholastic Inc.  2027-10-29        2057-04-28   
2  Harper Perennial Modern Classics  2061-09-16        2027-04-28   
3                    Modern Library  2039-05-08        2065-06-05   
4         Little, Brown and Company  2022-08-10        1974-06-15   

                                             premios  \
0  ['Locus Award Nominee for Best Young Adult Boo...   
1  ['Bram Stoker Award for Works for Young Reader...   
2  ['Pulitzer Prize for Fiction (1961)', 'Audie A...   
3                          Real worry two according.   
4  ['Georgia Peach Book Award (2007)', 'Buxtehude...   

                                        ambientacion  
0  ['District 12, Panem', 'Capitol, Panem', 'Pane...  
1  ['Hogwarts School of Witchcraft and Wizardry (...  
2               ['Maycomb, Alabama (United States)']  
3  ['United Kingdom', 'Derbyshire, England (Unite...  
4  ['Forks, Washington (United States)', 'Phoenix...

In [43]:
Valoracion.head()

bookid  calificacion  \
0                     2767052-the-hunger-games          4.33   
1  2.Harry_Potter_and_the_Order_of_the_Phoenix          4.50   
2                   2657.To_Kill_a_Mockingbird          4.28   
3                     1885.Pride_and_Prejudice          4.26   
4                               41865.Twilight          3.60   

   num_calificaciones                                          estrellas  \
0             6376780  ['3444695', '1921313', '745221', '171994', '93...   
1             2507623  ['1593642', '637516', '222366', '39573', '14526']   
2             4501075  ['2363896', '1333153', '573280', '149952', '80...   
3             2998241  ['1617567', '816659', '373311', '113934', '767...   
4             4964519  ['1751460', '1113682', '1008686', '542017', '5...   

   porcentaje_likes  puntuacion_bbe  votos_bbe  
0                96         2993816      30516  
1                98         2632233      26923  
2                95         2269402      23328  
3                94         1983116      20452  
4                78         1459448      14874

In [44]:
Autor.head()

nombre  num_trabajos  genero nacimiento     muerte  \
0   lexy v. golden            30  female 1938-10-11 2013-02-14   
1  جهاد عبد الحميد             1    male 1934-05-06 1982-04-01   
2    keith roberts           118    male 1935-09-20 2000-10-05   
3   sean fay wolfe            25    male 1958-08-23 2017-10-26   
4     karren brady             6  female 1930-08-05 1970-05-16   

   puntuacion_prom  num_votos            pais  
0             4.05        514      Montenegro  
1             3.52        591         Eritrea  
2             3.69       5402  United Kingdom  
3             4.61       1737         Hungary  
4             3.73        940       Hong Kong

In [45]:
print(f"Largo de la tabla Libro: {len(Libro)}")
print(f"Largo de la tabla Valoracion: {len(Valoracion)}")
print(f"Largo de la tabla Autor: {len(Autor)}")

Largo de la tabla Libro: 45429
Largo de la tabla Valoracion: 45429
Largo de la tabla Autor: 17910


### Largos Finales de las Tablas:

| **Tabla**     | **Largo**  |
|---------------|------------|
| Libro         | 45,429     |
| Valoracion    | 45,429     |
| Autor         | 17,910     |


In [46]:
# Prueba Final:
# Verificar unicidad de bookid en la tabla Valoracion
assert Valoracion['bookid'].is_unique, 'Hay llaves bookid duplicadas en la tabla Valoracion'

# Verificar unicidad de bookid en la tabla Libro
assert Libro['bookid'].is_unique, 'Hay llaves bookid duplicadas en la tabla Libro'

# Verificar unicidad de nombre en la tabla Autor
assert Autor['nombre'].is_unique, 'Hay llaves nombre duplicadas en la tabla Autor'

print("Todas las llaves son únicas en las tablas Valoracion, Libro y Autor.")

Todas las llaves son únicas en las tablas Valoracion, Libro y Autor.


# 6. Exportar Datos (.csv)

In [47]:
# Cambiar [] por {} solo en el inicio y final de los arreglos en Libro y Valoracion:
Libro['generos'] = Libro['generos'].str.replace(r'^\[', '{').str.replace(r'\]$', '}')
Libro['personajes'] = Libro['personajes'].str.replace(r'^\[', '{').str.replace(r'\]$', '}')
Libro['premios'] = Libro['premios'].str.replace(r'^\[', '{').str.replace(r'\]$', '}')
Libro['ambientacion'] = Libro['ambientacion'].str.replace(r'^\[', '{').str.replace(r'\]$', '}')
Valoracion['estrellas'] = Valoracion['estrellas'].str.replace(r'^\[', '{').str.replace(r'\]$', '}')

# Elimina los {} que no estan en los extremos de las columnas:
# Definir la función de limpieza
def limpiar_premios(premios):
    return re.sub(r'(?<!^)\{(.*?)\}(?!$)', r'\1', premios)

# Aplicar la función a la columna 'premios' del DataFrame 'Libro'
Libro['generos'] = Libro['generos'].apply(limpiar_premios)
Libro['personajes'] = Libro['personajes'].apply(limpiar_premios)
Libro['premios'] = Libro['premios'].apply(limpiar_premios)
Libro['ambientacion'] = Libro['ambientacion'].apply(limpiar_premios)
Valoracion['estrellas'] = Valoracion['estrellas'].apply(limpiar_premios)

# Añadir {} si corresponde
Libro['generos'] = Libro['generos'].apply(lambda x: '{' + x + '}' if not x.startswith('{') else x)
Libro['personajes'] = Libro['personajes'].apply(lambda x: '{' + x + '}' if not x.startswith('{') else x)
Libro['premios'] = Libro['premios'].apply(lambda x: '{' + x + '}' if not x.startswith('{') else x)
Libro['ambientacion'] = Libro['ambientacion'].apply(lambda x: '{' + x + '}' if not x.startswith('{') else x)
Valoracion['estrellas'] = Valoracion['estrellas'].apply(lambda x: '{' + x + '}' if not x.startswith('{') else x)

# Elimina todas las comillas simples de las cuatro columnas anteriores:
Libro['generos'] = Libro['generos'].str.replace("'", "")
Libro['personajes'] = Libro['personajes'].str.replace("'", "")
Libro['premios'] = Libro['premios'].str.replace("'", "")
Libro['ambientacion'] = Libro['ambientacion'].str.replace("'", "")
Valoracion['estrellas'] = Valoracion['estrellas'].str.replace("'", "")

# Eliminas las comas ',' que no tienen texto en su lado derecho:
Libro['generos'] = Libro['generos'].str.replace(',}', '}')
Libro['personajes'] = Libro['personajes'].str.replace(',}', '}')
Libro['premios'] = Libro['premios'].str.replace(',}', '}')
Libro['ambientacion'] = Libro['ambientacion'].str.replace(',}', '}')
Valoracion['estrellas'] = Valoracion['estrellas'].str.replace(',}', '}')

# Reemplaza las comas ',,' por ',' en las cuatro columnas anteriores:
Libro['generos'] = Libro['generos'].str.replace(',,', ',')
Libro['personajes'] = Libro['personajes'].str.replace(',,', ',')
Libro['premios'] = Libro['premios'].str.replace(',,', ',')
Libro['ambientacion'] = Libro['ambientacion'].str.replace(',,', ',')

In [48]:
# Buscar 'French-American' en algún elemento de algún array de la columna 'premios'
french_american_books = Libro[Libro['premios'].apply(lambda x: 'French-American' in x)]

# Mostrar los resultados
french_american_books['premios']

538      {[French-American Foundation Translation Prize...
682      {[BTBA Best Translated Book Award Nominee for ...
1529     {[Magnesia Litera for Translation (Litera za p...
2814     {[Prix Goncourt (2006), Independent Foreign Fi...
3111     {[Prix Médicis (1978), Mekka-prijs (1996), Fre...
3142     {[French-American Foundation Translation Prize...
7634     {[French-American Foundation Translation Prize...
10752    {[BTBA Best Translated Book Award Nominee for ...
11061    {[French-American Foundation Translation Prize...
12035    {[Grand Prix RTL-Lire (2014), Wellcome Book Pr...
28808    {[Prix Femina Vacaresco (1994), French-America...
28970    {[Prix Goncourt des Lycéens (2007), Independen...
33061    {[French-American Foundation Translation Prize...
39723    {[Prix Renaudot (2000), Prix Goncourt des Lycé...
45163    {[Prix Goncourt (1978), French-American Founda...
47567    {[French-American Foundation Translation Prize...
51150    {[BTBA Best Translated Book Award Nominee for .

In [54]:
# Buscar libros de George R.R. Martin en la columna 'autor'
george_rr_martin_books = Libro[Libro['autor'].str.contains("george r.r. martin", case=False, na=False)]

# Mostrar los resultados
george_rr_martin_books

bookid  \
46                        13496.A_Game_of_Thrones   
194                       62291.A_Storm_of_Swords   
248                        10572.A_Clash_of_Kings   
429                       13497.A_Feast_for_Crows   
471                 10664113-a-dance-with-dragons   
1259                     147915.A_Storm_of_Swords   
1578                     768889.A_Storm_of_Swords   
1679              12177850-a-song-of-ice-and-fire   
2899               9814682-a-song-of-ice-and-fire   
4322                          39943621-fire-blood   
4374                13337715-a-dance-with-dragons   
4510      18635622-a-knight-of-the-seven-kingdoms   
5039              13337716-a-dance-with-dragons-2   
5239                           382450.Fevre_Dream   
7098            8341706-die-herren-von-winterfell   
8178                 5739604-o-despertar-da-magia   
8265                  6383644-o-festim-dos-corvos   
8495                     67948.The_Armageddon_Rag   
9322                     67956.Dying_of_the_Light   
10381  5061675-a-game-of-thrones-a-clash-of-kings   
10526                   5739512-a-muralha-de-gelo   
11439                           147908.Wild_Cards   
12310                   11970747-the-hedge-knight   
13324                12111823-the-winds-of-winter   
16035                         776212.Tuf_Voyaging   
16220                      6859248-o-mar-de-ferro   
17303                   6382055-a-dream-of-spring   
22654                     1557943.Inside_Straight   
24745                        3347728-busted-flush   
24880                     1032784.Le_tr_ne_de_fer   
26693                       9550301-suicide-kings   
27003          22740195-the-world-of-ice-and-fire   
27237                        1282266.Hunter_s_Run   
28160                        13498.A_Song_for_Lya   
28470                            20519000-lowball   
29936                          9669552-fort-freak   
37822                  417729.Dreamsongs_Volume_I   
41588                            769902.Sandkings   
44823      32933122-das-lied-von-eis-und-feuer-10   
44988                             20168816-rogues   
47264   10222727-der-thron-der-sieben-k-nigreiche   
51059                        34896.The_Ice_Dragon   

                                                  titulo               autor  \
46                                     A Game of Thrones  george r.r. martin   
194                                    A Storm of Swords  george r.r. martin   
248                                     A Clash of Kings  george r.r. martin   
429                                    A Feast for Crows  george r.r. martin   
471                                 A Dance with Dragons  george r.r. martin   
1259                   A Storm of Swords: Blood and Gold  george r.r. martin   
1578                   A Storm of Swords: Steel and Snow  george r.r. martin   
1679                              A Song of Ice and Fire  george r.r. martin   
2899                              A Song of Ice and Fire  george r.r. martin   
4322                                        Fire & Blood  george r.r. martin   
4374               A Dance with Dragons: Dreams and Dust  george r.r. martin   
4510                      A Knight of the Seven Kingdoms  george r.r. martin   
5039             A Dance with Dragons 2: After the Feast  george r.r. martin   
5239                                         Fevre Dream  george r.r. martin   
7098                           Die Herren von Winterfell  george r.r. martin   
8178                                O Despertar da Magia  george r.r. martin   
8265                                 O Festim dos Corvos  george r.r. martin   
8495                                  The Armageddon Rag  george r.r. martin   
9322                                  Dying of the Light  george r.r. martin   
10381               A Game of Thrones / A Clash of Kings  george r.r. martin   
10526                                  A Muralha de Gelo  george r.r. martin   
114

In [50]:
Libro['personajes']
# Buscar "Scarlett O'Hara" en la columna 'personajes' de Libro
scarlett_ohara_books = Libro[Libro['personajes'].str.contains("Scarlett O'Hara", case=False, na=False)]

# Mostrar los resultados
scarlett_ohara_books

# Cambiar comillas dobles por comillas simples en los arreglos de la columna 'generos'
Libro['generos'] = Libro['generos'].str.replace('"', "'")

# Cambiar comillas dobles por comillas simples en los arreglos de la columna 'personajes'
Libro['personajes'] = Libro['personajes'].str.replace('"', "'")

# Cambiar comillas dobles por comillas simples en los arreglos de la columna 'premios'
Libro['premios'] = Libro['premios'].str.replace('"', "'")

# Cambiar comillas dobles por comillas simples en los arreglos de la columna 'ambientacion'
Libro['ambientacion'] = Libro['ambientacion'].str.replace('"', "'")

# Cambiar comillas dobles por comillas simples en los arreglos de la columna 'estrellas'
Valoracion['estrellas'] = Valoracion['estrellas'].str.replace('[', '{').str.replace(']', '}')

`Descomentar` para guardar las 3 tablas en formato `.csv`, en particular, recordar que cada ejecución completa del código con `runall` genera algunos datos nuevos por los creados a mano.

In [51]:
# Exportar la tabla Libro a CSV
Libro.to_csv('Libro.csv', index=False)

# Exportar la tabla Valoracion a CSV
Valoracion.to_csv('Valoracion.csv', index=False)

# Exportar la tabla Autor a CSV
Autor.to_csv('Autor.csv', index=False)

In [55]:
from collections import Counter

# Unir todos los premios en una sola lista
all_premios = Libro['premios'].str.cat(sep=',').replace('{', '').replace('}', '').split(',')

# Contar la frecuencia de cada premio
premio_counts = Counter(all_premios)

# Encontrar el premio con mayor frecuencia
most_common_premio = premio_counts.most_common(1)
print(f"El premio que aparece con mayor frecuencia es: {most_common_premio[0][0]} con {most_common_premio[0][1]} apariciones.")

El premio que aparece con mayor frecuencia es:  LovelyBooks Leserpreis Nominee for Allgemeine Literatur (2009)] con 27 apariciones.


In [56]:
from collections import Counter

# Unir todos los géneros en una sola lista
all_generos = Libro['generos'].str.cat(sep=',').replace('{', '').replace('}', '').split(',')

# Contar la frecuencia de cada género
genero_counts = Counter(all_generos)

# Encontrar los géneros más comunes
most_common_generos = genero_counts.most_common(10)
print("Los géneros más comunes son:")
for genero, count in most_common_generos:
    print(f"{genero.strip()}: {count} apariciones")

Los géneros más comunes son:
Fiction: 22955 apariciones
Romance: 10993 apariciones
Contemporary: 8868 apariciones
Fantasy: 8603 apariciones
Young Adult: 7738 apariciones
Adult: 6385 apariciones
Novels: 5771 apariciones
Audiobook: 5715 apariciones
[Fiction: 5639 apariciones
Historical: 5548 apariciones
